File: app/src/main/java/com/diplom/map/layers/utils/DbfStreamReader.java
package com.diplom.map.layers.utils;

import java.io.DataInputStream;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;

class DbfStreamReader extends DataInputStream {

    DataInputStream in;

    DbfStreamReader(InputStream in) {
        super(in);
        this.in = new DataInputStream(in);
    }

    String readString(int length) throws IOException {
        byte[] array = new byte[length];
        this.readFully(array);
        String s = new String(array,"windows-1251");
        return s.trim();
    }

    short readLEShort() throws IOException {
        int byte1 = this.in.read();
        int byte2 = this.in.read();
        if (byte2 == -1) {
            throw new EOFException();
        } else {
            return (short) ((byte2 << 8) + byte1);
        }
    }

    int readLEInt() throws IOException {
        int byte1;
        int byte2;
        int byte3;
        int byte4;
        synchronized (this) {
            byte1 = this.in.read();
            byte2 = this.in.read();
            byte3 = this.in.read();
            byte4 = this.in.read();
        }

        if (byte4 == -1) {
            throw new EOFException();
        } else {
            return (byte4 << 24) + (byte3 << 16) + (byte2 << 8) + byte1;
        }
    }

}

File: app/src/main/java/com/diplom/map/esri/model/DbfStreamReader.kt
package com.diplom.map.esri.model

import java.io.DataInputStream
import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

internal class DbfStreamReader(stream: InputStream) : DataInputStream(stream) {

    private val stream: DataInputStream = DataInputStream(stream)

    @Throws(IOException::class)
    fun readString(length: Int): String {
        val array = ByteArray(length)
        this.readFully(array)
        val s = String(array, StandardCharsets.UTF_8)
        return s.trim { it <= ' ' }
    }

    @Throws(IOException::class)
    fun readLEShort(): Short {
        val byte1 = this.stream.read()
        val byte2 = this.stream.read()
        return if (byte2 == -1) {
            throw EOFException()
        } else {
            ((byte2 shl 8) + byte1).toShort()
        }
    }

    @Throws(IOException::class)
    fun readLEInt(): Int {
        val byte1: Int
        val byte2: Int
        val byte3: Int
        val byte4: Int
        synchronized(this) {
            byte1 = this.stream.read()
            byte2 = this.stream.read()
            byte3 = this.stream.read()
            byte4 = this.stream.read()
        }

        return if (byte4 == -1) {
            throw EOFException()
        } else {
            (byte4 shl 24) + (byte3 shl 16) + (byte2 shl 8) + byte1
        }
    }

}
